### ISSUE TITLE
Issue with aliases

### ISSUE THREAD

**Issue by mboudet** [tripal/tripal#727]
_(Edited the initial issue to add all problems_
Hello,

I found several issues with the aliases functionality (and bulk_update feature).

1. Error in the tripal job when using the bulk aliases update:

When executing the job created by the bulk update feature, there is a php error.

```
Tripal Job Launcher
Running as user 'admin'
-------------------
2018-10-23 14:50:59: There are 1 jobs queued.
2018-10-23 14:50:59: Job ID 20.
2018-10-23 14:50:59: Calling: tripal_update_all_urls_and_titles(bio_data_1, organism/[taxrank__genus]/[taxrank__species], alias)
Undefined variable: token tripal.bulk_update.inc:25                                                                                                                                                            
```
The last commit in the  tripal.bulk_update.inc file seems to have changed variable names, and the token variable is not set ([here](https://github.com/tripal/tripal/blob/7.x-3.x/tripal/includes/tripal.bulk_update.inc#L25)). Reverting the last commit for this file seems to work.

2. Aliases are not functional until one is manually created:

If I set a redirect link for a type of entity, and then add some entities, the aliases are visible in the UI and the database. However, using the link redirect to a 404 error.
If I then manually add an alias, they all start working. It seems something is not active until the first alias creation.. 

3. Redirect link for entities are malformed:

In some cases , the alias created at entity creation will be missing some parts.
I have this issue with mrna

They link pattern is as follows :  _feature/[obi__organism,TAXRANK:0000005]/[obi__organism,TAXRANK:0000006]/[rdfs__type]/[data__identifier]_
An expected alias would be : 
feature/Citrus/sinensis/mRNA/PAC:18136223
But the created alias looks like this : 
feature///mRNA/
(So there are duplicate errors)

I added these materials programmatically, so the issue might come from there? It seems like at the entity creation, these fields are not available..

System info:

Tripal: 7.x-3.0
Drupal : 7.59
Postgres: 9.5.8
Php: 7.1.23

**Comment by laceysanderson** [comment]
Weird. I just checked this functionality again and it works for me. I also don't see any errors in my Tripal Job when bulk updating my URLs. Can you post what errors you received during the Tripal Job?

Also, can you try going to `bio_data/1` and see if it redirects to `organism/Drosophila/melanogaster`?

Your system information could also help us figure out what is going on:

* Tripal Version:
* Drupal Version:
* PostgreSQL Version:
* PHP Version:

Note: All information is available in your site's administrator report area (Administration Toolbar > Reports > Status Report

**Comment by mboudet** [comment]
Running the job gives me
```
Tripal Job Launcher
Running as user 'admin'
-------------------
2018-10-23 14:50:59: There are 1 jobs queued.
2018-10-23 14:50:59: Job ID 20.
2018-10-23 14:50:59: Calling: tripal_update_all_urls_and_titles(bio_data_1, organism/[taxrank__genus]/[taxrank__species], alias)
Undefined variable: token tripal.bulk_update.inc:25                                                                                                                                                            [notice]
2/2 entities have been updated. Memory usage: 34,647,872 bytes.
Done.Command dispatch complete
```
And the $token variable seems to unset since the last commit [here](https://github.com/tripal/tripal/blob/7.x-3.x/tripal/includes/tripal.bulk_update.inc#L25)
To bypass it, I reversed the last commit in this file, and it seemed to works, since the values are set in DB.

Going to `bio_data/1` just bring me to the organism page, without redirecting.

System info:

Tripal: 7.x-3.0
Drupal : 7.59
Postgres: 9.5.8
Php: 7.1.23

**Comment by mboudet** [comment]
Is there any module that needs to be activated for the url redirections to work?

Where is the redirection handled in the code? (Assuming the alias is properly set in DB)

**Comment by laceysanderson** [comment]
Redirection is handled by the Drupal Path module if memory serves me correctly. I'll be back to work from holiday on Wednesday and will try to help more this week.

**Comment by mboudet** [comment]
Ok, so after testing some more, aliases start working as soon as I manually create one (in admin/config/search/path).
Up until that point, they do not register. Maybe some variable needs to be set before they actually works?

**Comment by mboudet** [comment]
Additional tests : 
Calling drupal_clear_path_cache() without any content does not work. Even after adding content, the aliases are not working.

After adding an organism, and calling drupal_clear_path_cache(), the aliases are working, and any new content will have a functional alias.

As long as one alias is created before calling drupal_clear_path_cache(), it will works. I don't know why though.

**Comment by mboudet** [comment]
@laceysanderson I updated the first post for clarity, I actually have 3 separate issues (though some might be from drupal..).

**Comment by spficklin** [comment]
Hi @mboudet I ran through the same procedure you did, using the exact same token string that you did for the URLs and I was able to duplicate the problem.    Here's what I get when I try to use this URL:  `/feature/Citrus/sinensis/mRNA/PAC:18136223`

![image](https://user-images.githubusercontent.com/1719352/48636186-12c7c300-e97f-11e8-8e1f-2bc21a8ae608.png)

Can you confirm that this is the error you get?  If so, the problem is with the use of the word 'feature' in the URL token string.  In Tripal v2 we had a feature lookup tool that allows a site to provide a single lookup for any feature just by providing it's name. If more than one feature matches the name it will give the user a list of matching features and let them select from it.  We kept that functionality in Tripal v3 so as to keep Tripal as backwards compatible as possible.  The problem is that by using the word 'feature' as the first element of the URL Drupal is resolving that to our feature lookup tool rather than the alias you setup.   This is an unfortunate side-effect of keeping that functionality :-(

Could you recreate your aliases without 'feature' as the first word?  I tested replacing the word feature and it worked just fine.

I have made code changes to fix the bad code you pointed out and also added a warning about using 'feature' as the first word in the URL alias.  I have added pull request #753 to fix the problem.  @mboudet If you are comfortable testing the PR would you mind trying it out to see if the issue is resolved for you?

**Comment by mboudet** [comment]
That's interesting, it's seems like it might be another bug (regarding the feature part).
My issue with aliases not working is slightly different : aliases created automatically (using the pattern set in the entity setting) are not working until an alias is manually created ( via the /config/search/path tools, or with a script). Once the alias is created, all the other aliases start working..

![image](https://user-images.githubusercontent.com/17642511/48641349-bf369300-e9d9-11e8-8aeb-63a9eb71a5f7.png)

Here we can see the aliases set (and the bugged feature alias with missing tokens..)

Manually going to the page gives this : 

![image](https://user-images.githubusercontent.com/17642511/48641400-eee59b00-e9d9-11e8-8815-29bfb333de24.png)

If I now create manually an alias : 

![image](https://user-images.githubusercontent.com/17642511/48641467-28b6a180-e9da-11e8-985b-060b836f54e3.png)

The other aliases start working : 

![image](https://user-images.githubusercontent.com/17642511/48641497-4421ac80-e9da-11e8-999c-6531cf78d366.png)

Regarding the features, it's true that I had a different error (due to the issue you pointed out), but I could not see it since the aliases are malformed.
I usually upload data through a command line interface (python-tripal), so the issue might be coming from here? 

I'm going to try adding the features through the web interface and see if the pattern is working. 

Your update did fix the bulk update, thanks!

**Comment by spficklin** [comment]
So, this is a bit bizarre, I'm not able to reproduce the problem as I don't seem to encounter this. Once I do the bulk update of the URLs they just work. I don't have to manually add a URL via the GUI (or programmatically) to get it to work.... I will test again.

**Comment by mboudet** [comment]
Interesting. My process is this : launch a new tripal instance, edit the 'organism' tripal content to change the default link path, save it, and then add organism (through the web UI this time).
The alias is then properly generated, but non functional.

I'm deploying tripal trough docker, so the issue might come from a missing module in the docker image? I'm going to look into it.

**Comment by mboudet** [comment]
So, I looked into TripalEntityController.inc. When adding an entity, we setup the alias in the url_alias table, but we never call drupal_path_alias_whitelist_rebuild, so the alias list is not really set. (Though there are traces of it in comments).

When using the drupal function path_save (used when creating a new alias), this function is used after adding the values in DB.

Calling drupal_path_alias_whitelist_rebuild seems to solve the problem..  Though I have not idea why it works for you then.

Manually getting the path_alias_whitelist variable gives me an empty array if I do not call the drupal_path_alias_whitelist_rebuild function. (drush vget path_alias_whitelist)

Edit : It seems we stopped calling the drupal_path_alias_whitelist_rebuild function for performance issues.. it might be necessary to manually update the whitelist variable then.

**Comment by spficklin** [comment]
Hi @mboudet I tried again to replicate the problem but setting aliases always works with no problems.  But to hopefully help in your case, I have added the call to `drupal_path_alias_whitelist_rebuild()` function right after bulk updating of aliases.  It's in the branch named `727-tv3-bulk_aliases`.   Are you comfortable testing that branch to see if it fixes the issue?

**Comment by mboudet** [comment]
Hi, sorry for the delay! Thanks for your commit.
It did fix the bulk update not activating the aliases, but not the aliases not being functional before the bulk_update being used, due to drupal_path_alias_whitelist_rebuild not being called when an entity is created.
I added back drupal_path_alias_whitelist_rebuild [here](https://github.com/tripal/tripal/blob/7.x-3.x/tripal/includes/TripalEntityController.inc#L242), and it fixed my problem.

I'm still confused as to why I have the problem if it works for you. In any case, if adding drupal_path_alias_whitelist_rebuild back is too costly performance-wise, I can manage otherwise.

I still have an issue with the tokens being malformed. I'm going to look into it.

**Comment by spficklin** [comment]
We had to take out that `drupal_path_alias_whitelist_rebuild` because it dramatically slowed down bulk aliases. I wish we knew why we are seeing different behavior and why it's not working for you.  

I think perhaps, if we can't find out it may be better to close out this issue, especially since you have a work around.  If this pops up with someone else and we can see that it's happening in multiple places we can re-open. In the meantime, if you are having token problems perhaps that should be a new issue so we can get a clean start on it?  This thread is getting a bit long and hard for someone else to jump in and try to help. 

Let me know if you agree and I'll close this out.

**Comment by mboudet** [comment]
Sure. I think I managed to pinpoint the issue for the tokens, so I'll make another issue when I have a fix for it.

For the record, my workaround is manually creating an alias using the path_save function in a custom module I used.
Since drupal_path_alias_whitelist_rebuild needs to be called only once for the aliases to start working, it's enough.

**Comment by spficklin** [comment]
Okay.  Thanks!
